### PR TITLE
Add filter for elasticpress index name

### PR DESCRIPTION
**Pull request by hanifn**
## Description

Add a filter to rename ElasticPress indexes into something that would work more consistently in VIP Go.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).
- [x] This change has relevant documentation additions / updates (if applicable).

## Steps to Test

TBD

### CODE REVIEW

**Comment by mjangda** [comment]
In a separate PR, we should force disable Jetpack Search when our custom constant here is set.

File: vip-helpers/vip-elasticsearch.php [hunk 1]
```diff
@@ -450,4 +450,16 @@ function wpcom_search_api_wp_to_es_args( $args ) {
...
 
     return $es_query_args;
-}
\ No newline at end of file
+}
+
+/**
+ * Filter ElasticPress index name if using VIP ES infrastructure
+ */
+function filter_ep_index_name( $index_name, $blog_id, $indexables ) {
```
Comment (nickdaugherty) [inline vip-helpers/vip-elasticsearch.php]: Can we use a more unique function name? (or even an anonymous function).

File: vip-helpers/vip-elasticsearch.php [hunk 2]
```diff
@@ -450,4 +450,16 @@ function wpcom_search_api_wp_to_es_args( $args ) {
...
+	if ( defined( 'USE_VIP_ELASTICSEARCH' ) && true === USE_VIP_ELASTICSEARCH ) {
+		$index_name = 'vip-' . VIP_GO_APP_ID . '-' . $indexables->slug . '-' . $blog_id;
```
Comment (mjangda) [inline vip-helpers/vip-elasticsearch.php]: `sprintf` would improve the readbility of this line:

```suggestion
		$index_name = sprintf( 'vip-%s-%s-%s', VIP_GO_APP_ID, $indexables->slug, $blog_id );
```

File: vip-helpers/vip-elasticsearch.php [hunk 3]
```diff
@@ -450,4 +450,16 @@ function wpcom_search_api_wp_to_es_args( $args ) {
...
+	}
+
+	return $index_name;
+}
+add_filter( 'ep_index_name', 'filter_ep_index_name', PHP_INT_MAX, 3 );
```
Comment (mjangda) [inline vip-helpers/vip-elasticsearch.php]: As a best practice, any time a non-standard priority is used, it's helpful to include an inline comment noting why:

```suggestion
add_filter( 'ep_index_name', 'filter_ep_index_name', PHP_INT_MAX, 3 ); // We want to enforce the naming, so run this really late.
```

File: vip-helpers/vip-elasticsearch.php (hunk 1 above)
Comment (hanifn) [reply vip-helpers/vip-elasticsearch.php]: Changed in 7187875a860f2a742d2bbcac6cc4760300aecdfc

**Comment by hanifn** [comment]
Added some tests in e8afce86d371f37983fbf1eec7b925ebbf65585d

File: vip-helpers/vip-elasticsearch.php [hunk 4]
```diff
@@ -450,4 +450,16 @@ function wpcom_search_api_wp_to_es_args( $args ) {
...
-}
\ No newline at end of file
+}
+
+/**
+ * Filter ElasticPress index name if using VIP ES infrastructure
+ */
+function vip_elasticsearch_filter_ep_index_name( $index_name, $blog_id, $indexables ) {
+	if ( defined( 'USE_VIP_ELASTICSEARCH' ) && true === USE_VIP_ELASTICSEARCH ) {
+		$index_name = sprintf( 'vip-%s-%s-%s', VIP_GO_APP_ID, $indexables->slug, $blog_id );
```
Comment (simonwheatley) [inline vip-helpers/vip-elasticsearch.php]: We are populating `VIP_GO_APP_ID` with the client site ID. This contradicts what we consider to be the App ID in the Dashboard, etc, where we use the "Parent" (`production` environment) ID as the App ID for a group of environments.

Do we want to address this here, @mjangda @hanifn?

Comment (mjangda) [reply vip-helpers/vip-elasticsearch.php]: Yeah, good thinking. Problem is that we don't have a `VIP_GO_ENV_ID` exposed currently so we'll need to add that.

But, I think, for now, better to use `FILES_CLIENT_SITE_ID` instead.

File: tests/vip-helpers/test-vip-elasticsearch.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	/**
+	 * Test `ep_index_name` filter for ElasticPress + VIP ElasticSearch
+	 *
+	 * @runInSeparateProcess
+	 * @preserveGlobalState disabled
+	 */
+	public function test__vip_elasticsearch_filter_ep_index_name() {
+		$mock_indexable = (object) [ 'slug' => 'slug' ];
+
+		define( 'VIP_GO_APP_ID', 123 );
```
Comment (mjangda) [inline tests/vip-helpers/test-vip-elasticsearch.php]:
```suggestion
		define( 'FILES_CLIENT_SITE_ID', 123 );
```

File: tests/vip-helpers/test-vip-elasticsearch.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	 *
+	 * USE_VIP_ELASTICSEARCH not defined
+	 *
+	 * @runInSeparateProcess
+	 * @preserveGlobalState disabled
+	 */
+	public function test__vip_elasticsearch_filter_ep_index_name__no_constant() {
+		$mock_indexable = (object) [ 'slug' => 'slug' ];
+
+		define( 'VIP_GO_APP_ID', 123 );
```
Comment (mjangda) [inline tests/vip-helpers/test-vip-elasticsearch.php]:
```suggestion
		define( 'FILES_CLIENT_SITE_ID', 123 );
```

File: tests/vip-helpers/test-vip-elasticsearch.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	 *
+	 * USE_VIP_ELASTICSEARCH is false
+	 *
+	 * @runInSeparateProcess
+	 * @preserveGlobalState disabled
+	 */
+	public function test__vip_elasticsearch_filter_ep_index_name__constant_is_false() {
+		$mock_indexable = (object) [ 'slug' => 'slug' ];
+
+		define( 'VIP_GO_APP_ID', 123 );
```
Comment (mjangda) [inline tests/vip-helpers/test-vip-elasticsearch.php]:
```suggestion
		define( 'FILES_CLIENT_SITE_ID', 123 );
```

File: vip-helpers/vip-elasticsearch.php (hunk 4 above)
Comment (mjangda) [inline vip-helpers/vip-elasticsearch.php]:
```suggestion
		$index_name = sprintf( 'vip-%s-%s-%s', FILES_CLIENT_SITE_ID, $indexables->slug, $blog_id );
```

**Comment by nickdaugherty** [comment]
@mjangda out of curiosity, what's the benefit to using `FILES_CLIENT_SITE_ID` instead of `VIP_GO_APP_ID`?

**Comment by mjangda** [comment]
@nickdaugherty It's incorrectly named; see https://github.com/Automattic/vip-go-mu-plugins/pull/1374#discussion_r358157820

File: tests/vip-helpers/test-vip-elasticsearch.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	public function test__vip_elasticsearch_filter_ep_index_name__no_constant() {
+		$mock_indexable = (object) [ 'slug' => 'slug' ];
+
+		$index_name = apply_filters( 'ep_index_name', 'index-name', 1, $mock_indexable );
+
+		$this->assertEquals( 'index-name', $index_name );
+	}
+
+	/**
+	 * Test `ep_index_name` filter for ElasticPress + VIP ElasticSearch
```
Comment (nickdaugherty) [inline tests/vip-helpers/test-vip-elasticsearch.php]:
```suggestion
	 * Test `ep_index_name` filter for ElasticPress + VIP Elasticsearch
```

`_lowercase_s_dangit()`

File: tests/vip-helpers/test-vip-elasticsearch.php [hunk 9]
```diff
@@ -0,0 +1,53 @@
+<?php
+
+class VIP_ElasticSearch_Test extends \WP_UnitTestCase {
+	/**
+	 * Test `ep_index_name` filter for ElasticPress + VIP ElasticSearch
```
Comment (nickdaugherty) [inline tests/vip-helpers/test-vip-elasticsearch.php]:
```suggestion
	 * Test `ep_index_name` filter for ElasticPress + VIP Elasticsearch
```

File: tests/vip-helpers/test-vip-elasticsearch.php [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+		define( 'USE_VIP_ELASTICSEARCH', true );
+
+		$index_name = apply_filters( 'ep_index_name', 'index-name', 1, $mock_indexable );
+
+		$this->assertEquals( 'vip-123-slug-1', $index_name );
+	}
+
+	/**
+	 * Test `ep_index_name` filter for ElasticPress + VIP ElasticSearch
```
Comment (nickdaugherty) [inline tests/vip-helpers/test-vip-elasticsearch.php]:
```suggestion
	 * Test `ep_index_name` filter for ElasticPress + VIP Elasticsearch
```

**Comment by nickdaugherty** [comment]
r712-stacks

**Pull request merged**